File: Quest/Assets/Src/Cards/Abilities/Ability.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public abstract class Ability {
    
    private string name;
    private AdventureCard sourceCard;

    public Ability(string name, AdventureCard sourceCard)
    {
        this.name = name;
        this.sourceCard = sourceCard;
    }

    public abstract bool CanUseAbility(GameState gState, Player sourcePlayer);
    public abstract void UseAbility(GameState gState, Player sourcePlayer);

    public virtual bool DoesTarget(GameState gState) { return false; }

    public virtual bool CanTarget(GameState state, Targetable obj) { return false; }
    public virtual void SetTarget(GameState state, Targetable obj) { }

    public string GetName() { return name; }
    public AdventureCard getSourceCard() { return sourceCard; }
}

File: eclipse/Quest/src/model/cards/abilities/Ability.java
package model.cards.abilities;

import model.cards.AdventureCard;
import model.game.GameState;
import model.player.Player;

public abstract class Ability {
	private String name;
    private AdventureCard sourceCard;

    public Ability(String name, AdventureCard sourceCard)
    {
        this.name = name;
        this.sourceCard = sourceCard;
    }

    public abstract boolean CanUseAbility(GameState gState, Player sourcePlayer);
    public abstract void UseAbility(GameState gState, Player sourcePlayer);

    public boolean DoesTarget(GameState gState) { return false; }

    public boolean CanTarget(GameState state, Object obj) { return false; }
    public void SetTarget(GameState state, Object obj) { }

    public String GetName() { return name; }
    public AdventureCard getSourceCard() { return sourceCard; }
}
